docs/extending-entry-points.rst

```diff
@@ -0,0 +1,224 @@
+                    password=sftp_password,
+                    port=(parsed_url.port or 22),
+                    cnopts=cnopts,
+                ) as sftp:
+                    sftp.get(parsed_url.path, str(tmp_file))
+                return tmp_file.read_text()
+
+        @classmethod
+        def write(cls, file_content, path):
+            """
+            Write a notebook to an SFTP server.
+            """
+            parsed_url = urllib.parse.urlparse(path)
+            with tempfile.TemporaryDirectory() as tmpdir:
+                tmp_file = pathlib.Path(tmpdir) / "output.ipynb"
+                tmp_file.write_text(file_content)
+                with pysftp.Connection(
+                    parsed_url.hostname,
+                    username=sftp_username,
+                    password=sftp_password,
+                    port=(parsed_url.port or 22),
+                    cnopts=cnopts,
+                ) as sftp:
+                    sftp.put(str(tmp_file), parsed_url.path)
+
+        @classmethod
+        def pretty_path(cls, path):
+            return path
+
+        @classmethod
+        def listdir(cls, path):
+            raise NotImplementedError
+
+
+The ``setup.py`` file contains the following code:
+
+.. code-block:: python
+
+    from setuptools import setup, find_packages
+
+    setup(
+        name="papermill_sftp",
+        version="0.1",
+        url="https://github.com/my_username/papermill_sftp.git",
+        author="My Name",
+        author_email="my.email@gmail.com",
+        description="An SFTP I/O handler for papermill.",
+        packages=find_packages("./src"),
+        package_dir={"": "src"},
+        install_requires=["pysftp"],
+        entry_points={"papermill.io": ["sftp://=papermill_sftp:SFTPHandler"]},
+    )
+
+New engines
+-----------
+
+A papermill engine is a python object that can run, or execute, a notebook. The
+default implementation in papermill for example takes in a notebook object, and
+runs it locally on your machine.
+
+Developing a new engine
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Papermill engines need to inherit from the ``papermill.engines.Engine`` class.
+
+In order to be used, the new class needs to implement the class method
+``execute_managed_notebook``. The call signature should match that of the parent
+class:
+
+.. code-block:: python
+
+    class CustomEngine(papermill.engines.Engine):
+
+        @classmethod
+        execute_managed_notebook(cls, nb_man, kernel_name, **kwargs):
+            pass
+
+``nb_man`` is a |nbformat.NotebookNode|_, and ``kernel_name`` is a string. Your
+custom class then needs to implement the execution of the notebook. For example,
+you could insert code that executes the notebook remotely on a server, or
+executes the notebook many times to simulate different conditions.
+
+Ensuring your engine is found by papermill
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Much like I/O handlers, engines can be specified as entry-points, under the
```

Comment:
I'd write this assuming the reader arrives here without having first read the I/O handler section. Another pet peeve of mine is docs that assume you've read all of it when you are trying to use it as a reference.

Reply:
Yep - I've amended this.